File: libpkgupd/builders/meson.cc
#include "meson.hh"

#include "../exec.hh"
#include "../recipe.hh"

namespace rlxos::libpkgupd {
bool Meson::compile(Recipe const& recipe, std::string dir, std::string destdir,
                    std::vector<std::string> const& environ) {
  auto configure = recipe.configure();
  if (configure.find("--prefix") != std::string::npos) {
    configure = " --prefix=" + PREFIX + " --sysconfdir=" + SYSCONF_DIR +
                " --libdir=" + LIBDIR + " --libexecdir=" + LIBEXEDIR +
                " --bindir=" + BINDIR + " --sbindir=" + SBINDIR +
                " --datadir=" + DATADIR + " --localstatedir=" + CACHEDIR + " " +
                configure;
  }
  if (int status = Executor().execute(
          "meson " + configure + " _pkgupd_builddir", dir, environ);
      status != 0) {
    p_Error = "failed to do meson configuration";
    return false;
  }

  if (int status = Executor().execute(
          "ninja -C _pkgupd_builddir " + recipe.compile(), dir, environ);
      status != 0) {
    p_Error = "failed to build with meson";
    return false;
  }

  auto install = recipe.install();
  if (install.find("install ") == std::string::npos) {
    install = "install " + install;
  }

  if (int status = Executor().execute("ninja -C _pkgupd_builddir " + install,
                                      dir, environ);
      status != 0) {
    p_Error = "failed to install with meson";
    return false;
  }

  return true;
}
}  // namespace rlxos::libpkgupd

File: builder/buildsystem/meson/meson.go
package meson

import (
	"fmt"
	"log"
	"os"
	"os/exec"
	"path"
	"runtime"
	"strings"

	"github.com/itsmanjeet/pkgupd/recipe"
	"github.com/itsmanjeet/pkgupd/utils"
)

type Meson struct {
}

func (m Meson) Compile(mod *recipe.Module, srcdir string, destdir string) error {
	execute := func(bin string, args ...string) error {
		log.Printf("Executing %s, %s\n", bin, args)
		cmd := exec.Command(bin, args...)
		cmd.Dir = path.Join(srcdir, mod.Dir)
		cmd.Stderr = os.Stderr
		cmd.Stdin = os.Stdin
		cmd.Stdout = os.Stdout
		cmd.Env = os.Environ()
		cmd.Env = append(cmd.Env, "pkgupd_srcdir="+cmd.Dir)
		return cmd.Run()
	}

	configure := "meson"
	configure_args := []string{"--prefix=/usr", "--sysconfdir=/etc", "--libdir=/usr/lib", "--libexecdir=/usr/lib", "--sbindir=/usr/bin", "--bindir=/usr/bin", "--datadir=/usr/share", "--localstatedir=/var", utils.BUILD_DIR}
	if strings.Contains(mod.Configure, "--prefix") {
		configure_args = []string{utils.BUILD_DIR}
	}
	configure_args = append(configure_args, strings.Split(strings.Trim(os.ExpandEnv(mod.Configure), " "), " ")...)

	if err := execute(configure, configure_args...); err != nil {
		return fmt.Errorf("failed to configure %s", err)
	}
	compile_args := []string{"-C", utils.BUILD_DIR, fmt.Sprintf("-j%d", runtime.NumCPU())}
	if len(mod.Compile) != 0 {
		compile_user_args := strings.Split(strings.Trim(os.ExpandEnv(mod.Compile), " "), " ")
		if len(compile_user_args) != 0 {
			compile_args = append(compile_args, compile_user_args...)
		}

	}

	fmt.Printf("compile_args: -%s-\n", strings.Join(compile_args, "::"))

	if err := execute("ninja", compile_args...); err != nil {
		return fmt.Errorf("failed to compile %s", err)
	}

	install_args := []string{"-C", utils.BUILD_DIR, "install"}
	if strings.Contains(mod.Install, "install") {
		install_args = []string{}
	}

	if len(mod.Install) != 0 {
		install_args = append(install_args, strings.Split(strings.Trim(os.ExpandEnv(mod.Install), " "), " ")...)
	}

	if err := execute("ninja", install_args...); err != nil {
		return fmt.Errorf("failed to install %s", err)
	}

	return nil
}
